### PR TITLE
Format dicom

### PR DESCRIPTION
**Pull request by tcpan**
Updated read and write.  Better standard compliance for write.  Reading now allows random data access.

### CODE REVIEW

File: waveform_benchmark/formats/dicom.py [hunk 1]
```diff
@@ -72,555 +66,530 @@
...
+        # one instance per chunk
+        
+        # one dicomdir per study?  or series?
+        studyInstanceUID = uid.generate_uid()
+        seriesInstanceUID = uid.generate_uid()
+        prefix, ext = os.path.splitext(path)
+        # import json
+
+        # ======== organize the waveform chunks  (tested)
+        print("INPUT pleth", waveforms['Pleth'])
```
Comment (briangow) [inline waveform_benchmark/formats/dicom.py:423]: Perhaps this line should get commented out.

File: waveform_benchmark/formats/dicom.py [hunk 2]
```diff
@@ -72,555 +66,530 @@
...
+# CardiacElectrophysiologyWaveform: EPS, 1-4, , <=20000, DCID 3011 “Electrophysiology Anatomic Location” , SS
+# ArterialPulseWaveform: HD, 1, 1, ? , <600, DCID 3004 “Arterial Pulse Waveform” , SB/SS
+# RespiratoryWaveform: RESP, 1, 1, ? , <100, DCID 3005 “Respiration Waveform”, SB/SS
+# ScalpEEGWaveform:  EEG, 1 (interruption as separate instances), 1-64, , unconstrained, DCID 3030 “EEG Lead”, SS/SL
+# ElectromyogramWaveform: EMG, unconstrained, 1-64, , unconstrained, DCID 3031 “Lead Location Near or in Muscle” or DCID 3032 “Lead Location Near Peripheral Nerve”, SS/SL
+# SleepEEGWaveform: EEG, unconstrained, 1-64, , unconstrained, DCID 3030 “EEG Lead” , SS/SL
+# MultichannelRespiratoryWaveform: RESP, >1, , unconstrained, DCID 3005 “Respiration Waveform” , SS/SL
+# 
+
+CHANNEL_TO_DICOM_IOD = {
```
Comment (briangow) [inline waveform_benchmark/formats/dicom.py:101]: Do we need to add the other channel types from the proposed suite of waveforms for benchmarking? See Waveform suite characterization here: https://github.com/chorus-ai/chorus_waveform/discussions/11#discussioncomment-9225909 .

**Comment by tompollard** [comment]
Thanks @tcpan. I think we should go ahead and merge, but we will have to do some restructuring afterwards because our current framework expects each module in the `formats` folder to represent a separate format.

@bemoody @briangow my suggestion would be either:

1. allowing formats to include utilities in a subfolder within `formats`. e.g. 

    [waveform_benchmark/formats/dicom.py](https://github.com/chorus-ai/chorus_waveform/pull/52/files#diff-3fc37d74baf3ec0fd66c1331ce38a5f46f952fffe44c312b2bbe14f7966f2a6a) -> leave in current location

    [waveform_benchmark/formats/dcm_rand_access.py](https://github.com/chorus-ai/chorus_waveform/pull/52/files#diff-70734f84d0af575b61f624eefa7d00db7689b4c5d6bb3b085091719716460fc9) -> `waveform_benchmark/formats/dicom/rand_access.py`

    [waveform_benchmark/formats/dcm_waveform_reader.py](https://github.com/chorus-ai/chorus_waveform/pull/52/files#diff-9579e4e9bc3beb162f6b5b7c3664765d6fc767d763fb432c54804b2b9ddf16fa) -> `waveform_benchmark/formats/dicom/waveform_reader.py`

    [waveform_benchmark/formats/dcm_waveform_writer.py](https://github.com/chorus-ai/chorus_waveform/pull/52/files#diff-75fc6c3f139ec475167275fab15ac6607c3879bf9fdc139ca07effdce7b28b23) -> 
`waveform_benchmark/formats/dicom/waveform_writer.py`

2. Store `rand_access.py`, `waveform_reader.py`, `waveform_writer.py` in a separate repository, which is then added as a dependency here. This might make sense if we consider the modules to be implementing the format, rather than just applying it.

**Review by briangow — COMMENTED** [review]
@tcpan , thanks so much for this - preparing the DICOM format for benchmarking clearly took a lot of work! 

I've left a couple of comments in the code. One other thought is that it might be cleaner to move the dcm_* files under a separate folder (perhaps formats/dcm_utils/dcm_*) such that the only files at the formats/ level are the ones that get called for benchmarking. This is completely up to you though, no worries if you prefer to leave them where they are.

**Comment by tcpan** [comment]
Hi, Brian, sorry for the delay.  Was working on house stuff all day.  Will do re subfolder. Is formats the right place to put that?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by briangow** [comment]
> Hi, Brian, sorry for the delay.  Was working on house stuff all day.  Will do re subfolder. Is formats the right place to put that?

@tcpan , it's up to you. If you feel that these files are useful beyond this repository perhaps they deserve a dedicated place as @tompollard suggested in this 2nd thought on the topic. It would probably make sense to create a general utils module in this repository at some point. If you prefer that option feel free to set it up or let me know and I'd be happy to do so. As far as I'm concerned, having them in a subfolder under formats/ is also fine.

File: waveform_benchmark/formats/dicom.py (hunk 1 above)
Comment (tcpan) [reply waveform_benchmark/formats/dicom.py]: done.

File: waveform_benchmark/formats/dicom.py (hunk 2 above)
Comment (tcpan) [reply waveform_benchmark/formats/dicom.py:101]: yes, this needs to be updated.

**Comment by tcpan** [comment]
I have checked in some updates, with some fixes and some new tests for different chunking strategy (likely not to make a huge difference).   Note that to get the channel metadata, each file is opened and scanned. A more efficient way to do this would be to add the channel metadata as a private tag on the DICOMDIR file.  This would still be standards compliant, but I have not implemented this.

**Comment by briangow** [comment]
Thanks @tcpan !

File: waveform_benchmark/formats/dicom.py (hunk 2 above)
Comment (briangow) [reply waveform_benchmark/formats/dicom.py:101]: @tcpan , I've started an issue to track this: https://github.com/chorus-ai/chorus_waveform/issues/53

**Comment by tompollard** [comment]
> Hi, Brian, sorry for the delay. Was working on house stuff all day. Will do re subfolder. Is formats the right place to put that?

>> @tcpan , it's up to you. If you feel that these files are useful beyond this repository perhaps they deserve a dedicated place as @tompollard suggested in this 2nd thought on the topic. It would probably make sense to create a general utils module in this repository at some point. If you prefer that option feel free to set it up or let me know and I'd be happy to do so. As far as I'm concerned, having them in a subfolder under formats/ is also fine.

I think the cleanest option is to store `rand_access.py`, `waveform_reader.py`, `waveform_writer.py` in a separate "DICOM-waveform" repository, because (as far as I understand it) the code implements the functionality for writing waveforms to DICOM. If we choose DICOM as the format, we will need to have a package that enables these functions (and actually, I assume there will be demand by the community for this functionality regardless). I'd be happy to help set up a pip installable version of this repo if you like this option.

If we want to go for the quick and easy option, my preference would be to move them to a submodule of the `formats` folder. I don't particularly like the idea of a generic `utils` submodule, because it has potential to become a dumping ground for random code.

**Pull request merged**